text/1180971/async_afek_gafni.tex

```diff
@@ -0,0 +1,181 @@
+    \item ASK - the message send by the ordinary process to candidate because he saw node with higher (level, id) than the ordinary node - he send it to the father to eliminate one node (father or sender). He waits for one of the three messages below (he doesn't process other messages in between - not including LEADER, END).
+    \item ANSWER\_ACCEPT\_CND - the message send by candidate process in response to ASK message. The candidate accepted his fate and was killed by the ASK message, he needs to inform the sender of ASK message.
+    \item ANSWER\_DENY\_CND - the message send by candidate process in response to ASK message. The candidate managed to defeat ASK message but need to response to the asker.
+    \item DEAD - the message send by candidate process in response to ASK message. However the node is already dead (ordinary process and candidate process runs until the end of program - dead candidate process only answers the ASK messages and don't send new ARRIVE messages).
+    \item LEADER - the leader was found! The node send that information to every node with his id. Every other node will now send END message to the leader and prepare to end the program.
+    \item END - if the node is not the leader then he sends this messages to the leader informing leader that he is ready to end the program (he will receive messages but it will discard them). If leader gets n - 1 END messages (from every neighbour) then he sends END message to all nodes and then the program ends. It is needed to have two kinds of messages because if there was only END message (if node receive it - ends) then there would be a possibility that some node would send on closed channel.
+\end{itemize}
+
+\section*{Implementation details}
+First of all algorithm above assume that channels are not blocking. However in our implementation the node can only send one message without blocking the channel. If it sends two to the same channel the node waits until the first message is read.
+
+So the first change we implement is that every ordinary process wait for response from candidate father. We need to change also candidate process, Now candidate on response to ASK message from ordinary process will return one of three values:
+\begin{enumerate}
+    \item DEAD 
+    \item ANSWER\_ACCEPT\_CND 
+    \item ANSWER\_DENY\_CND
+\end{enumerate}
+
+\section*{Proof of correctness}
+The pair (level, id) is always increasing so there will always be the maximum candidate.
+
+\section*{Proof of complexity}
+\begin{lemma}
+For any given \(k\), the number of candidates that own \(\frac{n}{k}\) or more nodes is at most \(k\). 
+\end{lemma}
+\begin{proof}
+Let \(C_1\) and \(C_2\) be any two candidates which owned \(\frac{n}{k}\) nodes at some point in time. \(C1\) and \(C2\) must have owned disjoint sets of at least \(\frac{n}{k}\) nodes each. If they never tried to claim a node from each other, we are
```

Comment:
Please replace C1 by C_1 and C2 by C_2 whenever it occurs.

Reply:
Done